File: analyzer/src/mlview/core/pipelines.py
"""MLV-P12: weakly-connected pipelines seeded from `workspace.entrypoints`.

CONTRACTS 11.47. A research repo with ten training scripts is ten pipelines, not
one 320-node graph, and `workspace.entrypoints` already knows their names. This
module turns that into a **relation over the finished document** - pure, and
readable by both ports from the document alone:

* `seeds(E)`  = the nodes whose `loc.file` is the entrypoint `E`;
* adjacency   = `data` and `call` edges, both directions, **plus containment**
  (`parent` up and down). `config` and `control` edges are deliberately cut: a
  shared `config.py` is precisely the module that would merge ten independent
  training scripts into one component;
* `reach(E)`  = the closure of `seeds(E)`, which **includes but does not expand
  through** a node belonging to another entrypoint's file. Without that one
  clause an undirected closure is the whole connected component whichever seed
  it starts from, ten training scripts that share one `utils.py` are one
  pipeline, and the feature answers nothing;
* a node in `reach(E)` that another entrypoint also reaches is **shared** for
  `E` - unless it is one of `E`'s own seeds, which are always `E`'s. A node in
  no reach at all is **unreached** and belongs to no pipeline.

Nothing here reads the emitted `pipelines[]` block: `project()` recomputes it,
so a hand-edited or stale block can never move a projection. Everything is
materialised in **document order**, never set-iteration order, because both
ports must agree byte for byte.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Sequence, Set, Tuple

from .selectors import ascii_lower

__all__ = ["PipelineIndex", "build_index", "pipelines_block", "pipeline_catalog",
           "resolve_entrypoint", "PIPELINE_EDGE_KINDS"]

#: The only edge kinds that join a pipeline (11.47 A2).
PIPELINE_EDGE_KINDS = ("data", "call")

_ZERO_COUNTS = {"low": 0, "medium": 0, "high": 0}
_SEVERITIES = ("high", "medium", "low")


def _max_severity(counts: Dict[str, int]) -> Optional[str]:
    return next((s for s in _SEVERITIES if counts.get(s)), None)


@dataclass(frozen=True)
class PipelineIndex:
    """Every pipeline of one document, computed once (11.47 A)."""

    entrypoints: Tuple[str, ...] = ()
    #: entrypoint -> its reachable node ids, in document order.
    reach: Dict[str, Tuple[str, ...]] = field(default_factory=dict)
    #: node ids reached by two or more entrypoints.
    shared: frozenset = frozenset()
    #: node id -> the entrypoint whose file it lives in, for the seeds only.
    owner: Dict[str, str] = field(default_factory=dict)
    #: node ids reachable from none.
    unreached: Tuple[str, ...] = ()

    def _shared_for(self, entrypoint: str, node_id: str) -> bool:
        """A node another entrypoint also reaches - unless it is one of THIS
        entrypoint's own seeds, which are never taken away from it."""
        return (node_id in self.shared
                and self.owner.get(node_id) != entrypoint)

    def core_of(self, entrypoint: str) -> Tuple[str, ...]:
        """The nodes this pipeline does **not** share with another (11.47 C)."""
        return tuple(n for n in self.reach.get(entrypoint, ())
                     if not self._shared_for(entrypoint, n))

    def context_of(self, entrypoint: str) -> Tuple[str, ...]:
        """The shared nodes of this pipeline - forced to `viewRole: context`."""
        return tuple(n for n in self.reach.get(entrypoint, ())
                     if self._shared_for(entrypoint, n))

    @property
    def non_empty(self) -> Tuple[str, ...]:
        return tuple(e for e in self.entrypoints if self.reach.get(e))


def _adjacency(nodes: Sequence[Dict[str, Any]],
               edges: Sequence[Dict[str, Any]]) -> Dict[str, List[str]]:
    """`data` + `call` edges, both directions, plus containment (11.47 A2)."""
    known = {n.get("id") for n in nodes}
    adj: Dict[str, List[str]] = {}

    def link(a: Optional[str], b: Optional[str]) -> None:
        if not a or not b or a == b or a not in known or b not in known:
            return
        adj.setdefault(a, []).append(b)
        adj.setdefault(b, []).append(a)

    for edge in edges:
        if edge.get("kind") in PIPELINE_EDGE_KINDS:
            link(edge.get("source"), edge.get("target"))
    for node in nodes:
        link(node.get("id"), node.get("parent"))
    return adj


def build_index(graph: Dict[str, Any]) -> PipelineIndex:
    """Compute every pipeline of `graph`. Pure; O(entrypoints x (V + E))."""
    nodes: List[Dict[str, Any]] = list(graph.get("nodes") or [])
    edges: List[Dict[str, Any]] = list(graph.get("edges") or [])
    workspace = graph.get("workspace") or {}
    entrypoints: Tuple[str, ...] = tuple(
        str(e) for e in (workspace.get("entrypoints") or []))
    if not entrypoints or not nodes:
        return PipelineIndex(entrypoints=entrypoints,
                             unreached=tuple(n["id"] for n in nodes))

    adj = _adjacency(nodes, edges)
    order = {n["id"]: i for i, n in enumerate(nodes)}
    by_file: Dict[str, List[str]] = {}
    for node in nodes:
        by_file.setdefault((node.get("loc") or {}).get("file") or "", []).append(
            node["id"])

    owner: Dict[str, str] = {}
    for entrypoint in entrypoints:
        for node_id in by_file.get(entrypoint, ()):
            owner.setdefault(node_id, entrypoint)

    reach: Dict[str, Tuple[str, ...]] = {}
    hits: Dict[str, int] = {}
    for entrypoint in entrypoints:
        seen: Set[str] = set()
        stack = list(by_file.get(entrypoint, ()))
        while stack:
            current = stack.pop()
            if current in seen:
                continue
            seen.add(current)
            held = owner.get(current)
            if held is not None and held != entrypoint:
                continue            # another entrypoint's own node: keep, stop
            stack.extend(adj.get(current, ()))
        if not seen:
            reach[entrypoint] = ()
            continue
        reach[entrypoint] = tuple(sorted(seen, key=lambda n: order.get(n, 0)))
        for node_id in seen:
            hits[node_id] = hits.get(node_id, 0) + 1

    shared = frozenset(n for n, count in hits.items() if count > 1)
    unreached = tuple(n["id"] for n in nodes if n["id"] not in hits)
    return PipelineIndex(entrypoints=entrypoints, reach=reach, shared=shared,
                         owner=owner, unreached=unreached)


def pipelines_block(graph: Dict[str, Any]) -> List[Dict[str, Any]]:
    """The optional root `pipelines[]` rows, in `workspace.entrypoints` order.

    One row per **non-empty** pipeline. The caller decides whether to emit the
    block at all: 11.47 D emits it only at two or more rows, because the block
    exists to drive a chooser and a single-pipeline workspace has nothing to
    choose.

    `issueCounts` counts non-suppressed issues with at least one anchor inside
    the pipeline, so an issue spanning two pipelines is counted in **both**: the
    block is a menu, not a partition, and the rows are deliberately not required
    to sum to `stats.issues`.
    """
    index = build_index(graph)
    issues = [i for i in graph.get("issues") or [] if not i.get("suppressed")]
    rows: List[Dict[str, Any]] = []
    for entrypoint in index.entrypoints:
        members = index.reach.get(entrypoint) or ()
        if not members:
            continue
        inside = set(members)
        counts = dict(_ZERO_COUNTS)
        for issue in issues:
            if any(n in inside for n in issue.get("nodeIds") or []):
                severity = issue.get("severity")
                if severity in counts:
                    counts[severity] += 1
        shared = len(index.context_of(entrypoint))
        rows.append({
            "entrypoint": entrypoint,
            "label": entrypoint,
            "nodeCount": len(members),
            "exclusiveCount": len(members) - shared,
            "sharedCount": shared,
            "issueCounts": counts,
        })
    return rows


def resolve_entrypoint(graph: Dict[str, Any], target: str
                       ) -> Tuple[Optional[str], List[str]]:
    """Match `target` against `workspace.entrypoints` (11.47 B).

    Exact path, then bare basename when the target carries no `/`, then the same
    two ASCII-case-folded - the `file:` ladder, for the same reason. Returns
    `(canonical entrypoint or None, warnings)`.
    """
    entrypoints = [str(e) for e in
                   ((graph.get("workspace") or {}).get("entrypoints") or [])]
    text = (target or "").replace("\\", "/")
    if text in entrypoints:
        return text, []
    if "/" not in text and text:
        hits = [e for e in entrypoints if e.rsplit("/", 1)[-1] == text]
        if hits:
            return hits[0], []
    want = ascii_lower(text)
    folded = [e for e in entrypoints if ascii_lower(e) == want]
    if not folded and "/" not in text and text:
        folded = [e for e in entrypoints
                  if ascii_lower(e.rsplit("/", 1)[-1]) == want]
    if folded:
        return folded[0], [
            "scope pipeline:%s matched case-insensitively; the canonical "
            "spelling is %s" % (text, folded[0])]
    return None, []


def pipeline_catalog(graph: Dict[str, Any]) -> List[Dict[str, Any]]:
    """One `pipeline:` row per pipeline, in `workspace.entrypoints` order.

    Returned **only when the workspace has two or more non-empty pipelines**,
    for the same reason 11.47 D emits the document block only then: a single
    pipeline is not a choice. Rows carry exactly the ten keys `scope_catalog`
    rows carry, so `--list-scopes` renders both lists through one formatter.

    `nodeCount` keeps the column's documented meaning - *what `--scope <spec>
    --depth 0` draws as `core`* - which for a pipeline is its **exclusive**
    reach, not `reach(E)`; the shared remainder is drawn as context and is
    reported in the document's `pipelines[]` block as `sharedCount`. Deliberately
    NOT merged into `scope_catalog`: the MCP `mlview_graph {scope:"units"}`
    catalogue prefixes every row with `unit:` from its `qualname`
    (`claude-plugin/server/mlview_scope.py`), so a pipeline row there would
    render as a selector that does not resolve.
    """
    index = build_index(graph)
    if len(index.non_empty) < 2:
        return []
    nodes = list(graph.get("nodes") or [])
    issues = [i for i in graph.get("issues") or [] if not i.get("suppressed")]
    rows: List[Dict[str, Any]] = []
    for entrypoint in index.non_empty:
        core = index.core_of(entrypoint)
        inside = set(core)
        counts = dict(_ZERO_COUNTS)
        for issue in issues:
            if any(n in inside for n in issue.get("nodeIds") or []):
                severity = issue.get("severity")
                if severity in counts:
                    counts[severity] += 1
        seeds = [n["id"] for n in nodes
                 if (n.get("loc") or {}).get("file") == entrypoint]
        rows.append({
            "spec": "pipeline:%s" % entrypoint,
            "kind": "pipeline",
            "nodeId": seeds[0] if seeds else "",
            "label": entrypoint.rsplit("/", 1)[-1],
            "qualname": entrypoint,
            "file": entrypoint,
            "line": 1,
            "nodeCount": len(core),
            "issueCounts": counts,
            "maxSeverity": _max_severity(counts),
        })
    return rows

File: webview/src/scope/pipelines.ts
/**
 * MLV-P12 — the pipeline relation (CONTRACTS 11.47 A), ported line for line from
 * `analyzer/src/mlview/core/pipelines.py`.
 *
 * A research repo with ten training scripts renders as one graph at 22 % zoom.
 * The analyzer already knows there are ten entrypoints; what nobody could say
 * was *"the exp03 pipeline"*, which is the unit a practitioner thinks in. A
 * pipeline is therefore ANOTHER PROJECTION — one new selector kind, one optional
 * root block, no new mode — and this module computes the relation it projects
 * over.
 *
 * The relation, verbatim from 11.47 A:
 *
 *   A1  seeds(E) = the nodes whose `loc.file == E`, in DOCUMENT order.
 *   A2  adjacency = every `data` or `call` edge, in both directions, UNION the
 *       containment relation `parent` <-> child, in both directions. `config`
 *       and `control` edges are deliberately excluded: a shared `config.py` is
 *       precisely the module that would merge ten independent training scripts
 *       into one component, which is the failure mode the item exists to avoid.
 *   A3  reach(E) = the closure of seeds(E) under A2, which INCLUDES BUT DOES NOT
 *       EXPAND THROUGH a node that belongs to another entrypoint's own file.
 *       Without that one clause an undirected closure is the whole connected
 *       component whichever seed it starts from — ten training scripts sharing
 *       one `utils.py` would be one pipeline and the feature would answer
 *       nothing. A node another entrypoint also reaches is SHARED for E, unless
 *       it is one of E's OWN seeds, which are never taken away from it. A node
 *       in no reach at all is UNREACHED, which is a finding about the workspace
 *       rather than an error.
 *   A4  every set is materialised in DOCUMENT order, never in set-iteration
 *       order — the same determinism rule the projection lives by (11.2.1).
 *
 * **Both ports compute this; neither trusts the emitted `pipelines[]` block.**
 * The block is a menu the analyzer wrote for a chooser; the projection is
 * computed here, so a stale or absent block can never change what is drawn.
 * `pipelineDrift()` is the one place the two are compared, and it REPORTS a
 * disagreement rather than resolving it.
 *
 * Pure: no DOM, no clock, no set-iteration-order leak.
 */

import { ScopeError, asciiLower } from './selector.js';
import type { ScopeResolution } from './project.js';
import type { Scope } from './selector.js';
import type { IssueCounts, MLGraph, MLNode, Severity } from '../types.js';

/** One row of the chooser and of the scope picker's Pipelines section. */
export interface PipelineRow {
  /** The canonical `workspace.entrypoints` path — what `pipeline:` takes. */
  entrypoint: string;
  /** `|reach(E)|`. */
  nodeCount: number;
  /** Reachable from this entrypoint and no other — the projection's `core`. */
  exclusiveCount: number;
  /** Reachable from another entrypoint too — drawn as `context` (11.47 C). */
  sharedCount: number;
  /** Non-suppressed findings with at least one anchor in `reach(E)`. */
  issueCounts: IssueCounts;
}

const SEVERITIES: Severity[] = ['high', 'medium', 'low'];

/** The relation over one document. Built once, asked many times. */
export class PipelineIndex {
  /** `workspace.entrypoints`, in the analyzer's ranked order (11.47 D). */
  readonly entrypoints: string[];
  /** entrypoint -> reach(E), in document order. */
  private readonly reachOf = new Map<string, string[]>();
  /** Nodes reached by two or more entrypoints. */
  readonly shared: Set<string>;
  /** node id -> the entrypoint whose file it lives in, for the SEEDS only. */
  private readonly owner = new Map<string, string>();
  /** Nodes in no reach at all, in document order. */
  readonly unreached: string[];
  private readonly graph: MLGraph;

  constructor(graph: MLGraph) {
    this.graph = graph;
    const nodes = graph.nodes || [];
    const adjacency = buildAdjacency(graph);
    const entrypoints = (graph.workspace && graph.workspace.entrypoints) || [];
    this.entrypoints = entrypoints.slice();
    this.shared = new Set<string>();
    if (!this.entrypoints.length || !nodes.length) {
      this.unreached = nodes.map((n) => n.id);
      return;
    }

    // The seeds of every entrypoint, claimed first-come in entrypoint order.
    for (const entry of this.entrypoints) {
      for (const id of seedIds(nodes, entry)) {
        if (!this.owner.has(id)) this.owner.set(id, entry);
      }
    }

    const hits = new Map<string, number>();
    for (const entry of this.entrypoints) {
      const reached = closure(seedIds(nodes, entry), adjacency, entry, this.owner);
      if (!reached.size) {
        this.reachOf.set(entry, []);
        continue;
      }
      // A4: back to DOCUMENT order. A set is a membership test, never an order.
      const ordered = nodes.filter((n) => reached.has(n.id)).map((n) => n.id);
      this.reachOf.set(entry, ordered);
      for (const id of ordered) hits.set(id, (hits.get(id) || 0) + 1);
    }
    const unreached: string[] = [];
    for (const node of nodes) {
      const count = hits.get(node.id) || 0;
      if (count > 1) this.shared.add(node.id);
      else if (count === 0) unreached.push(node.id);
    }
    this.unreached = unreached;
  }

  /**
   * Shared FOR this entrypoint: another entrypoint reaches it too, and it is
   * not one of this entrypoint's own seeds — a script's own statements are
   * never taken away from it and handed to a neighbour as context.
   */
  private sharedFor(entrypoint: string, id: string): boolean {
    return this.shared.has(id) && this.owner.get(id) !== entrypoint;
  }

  /** `reach(E)` in document order, or `[]` for an entrypoint we do not have. */
  reach(entrypoint: string): string[] {
    return this.reachOf.get(entrypoint) || [];
  }

  /** A1: the nodes whose file IS the entrypoint, in document order. */
  seeds(entrypoint: string): string[] {
    return (this.graph.nodes || [])
      .filter((n) => fileOf(n) === entrypoint)
      .map((n) => n.id);
  }

  /** `core` for 11.47 C: reachable from here and not shared away. */
  exclusive(entrypoint: string): string[] {
    return this.reach(entrypoint).filter((id) => !this.sharedFor(entrypoint, id));
  }

  /** `forced context` for 11.47 C: reachable from here AND from elsewhere. */
  sharedWith(entrypoint: string): string[] {
    return this.reach(entrypoint).filter((id) => this.sharedFor(entrypoint, id));
  }

  /**
   * The chooser's rows, in `workspace.entrypoints` order — which 11.47 D says is
   * already ranked most-likely-first, so this never re-sorts.
   *
   * Entrypoints whose reach is EMPTY are dropped: 11.47 D emits the block "only
   * when the document has two or more non-empty pipelines", and a menu row that
   * draws nothing is a row that reads as a bug.
   */
  rows(): PipelineRow[] {
    const out: PipelineRow[] = [];
    for (const entry of this.entrypoints) {
      const reached = this.reach(entry);
      if (!reached.length) continue;
      // `sharedWith`, exactly as `pipelines_block` uses `context_of`: the row's
      // `sharedCount` is what THIS view would draw as context, so a script's own
      // statements — which another entrypoint may well reach — are never counted
      // against it. `exclusiveCount + sharedCount == nodeCount` either way; only
      // the split moves, and it moves to match the projection.
      const shared = this.sharedWith(entry).length;
      out.push({
        entrypoint: entry,
        nodeCount: reached.length,
        exclusiveCount: reached.length - shared,
        sharedCount: shared,
        issueCounts: countIssues(this.graph, new Set(reached)),
      });
    }
    return out;
  }
}

/** The only edge kinds that join a pipeline (11.47 A2). */
export const PIPELINE_EDGE_KINDS = ['data', 'call'];

/**
 * A2. `data` and `call` edges both ways, plus `parent` <-> child both ways.
 *
 * A link to an id the document does not contain, and a self-link, are both
 * dropped — exactly as `_adjacency` in `core/pipelines.py` drops them, so a
 * malformed document cannot make the two ports disagree about a reach.
 */
function buildAdjacency(graph: MLGraph): Map<string, string[]> {
  const known = new Set<string>((graph.nodes || []).map((n) => n.id));
  const adjacency = new Map<string, string[]>();
  const link = (a: string, b: string): void => {
    if (!a || !b || a === b || !known.has(a) || !known.has(b)) return;
    push(adjacency, a, b);
    push(adjacency, b, a);
  };
  for (const edge of graph.edges || []) {
    if (PIPELINE_EDGE_KINDS.indexOf(edge.kind) < 0) continue;
    link(edge.source, edge.target);
  }
  for (const node of graph.nodes || []) {
    link(node.id, node.parent || '');
  }
  return adjacency;
}

function push(map: Map<string, string[]>, key: string, value: string): void {
  const list = map.get(key);
  if (list) list.push(value);
  else map.set(key, [value]);
}

function seedIds(nodes: MLNode[], entrypoint: string): string[] {
  return nodes.filter((n) => fileOf(n) === entrypoint).map((n) => n.id);
}

/**
 * A3: the closure, cycle-guarded and bounded by the node set.
 *
 * The one clause that makes a pipeline a pipeline: a node that belongs to
 * ANOTHER entrypoint's own file is KEPT — it really is reachable from here — but
 * the walk does not expand through it. Without it an undirected closure is the
 * whole connected component whichever seed it starts from, and ten training
 * scripts sharing one `utils.py` are one pipeline.
 */
function closure(
  seeds: string[],
  adjacency: Map<string, string[]>,
  entrypoint: string,
  owner: Map<string, string>,
): Set<string> {
  const seen = new Set<string>();
  const stack = seeds.slice();
  while (stack.length) {
    const current = stack.pop() as string;
    if (seen.has(current)) continue;
    seen.add(current);
    const held = owner.get(current);
    if (held !== undefined && held !== entrypoint) continue; // keep, but stop
    for (const next of adjacency.get(current) || []) {
      if (!seen.has(next)) stack.push(next);
    }
  }
  return seen;
}

function fileOf(node: MLNode): string {
  return (node.loc && node.loc.file) || '';
}

function basename(path: string): string {
  const at = path.lastIndexOf('/');
  return at < 0 ? path : path.slice(at + 1);
}

/** 11.47 D: non-suppressed findings with at least one anchor in the set. */
function countIssues(graph: MLGraph, ids: Set<string>): IssueCounts {
  const counts: IssueCounts = { low: 0, medium: 0, high: 0 };
  for (const issue of graph.issues || []) {
    if (issue.suppressed) continue;
    if (!(issue.nodeIds || []).some((id) => ids.has(id))) continue;
    const severity = issue.severity as Severity;
    if (counts[severity] !== undefined) counts[severity]++;
  }
  return counts;
}

/** The highest severity among a row's findings, for the picker's `data-sev`. */
export function rowSeverity(row: PipelineRow): Severity | null {
  return SEVERITIES.filter((s) => row.issueCounts[s] > 0)[0] || null;
}

/**
 * Resolve a `pipeline:` target to a canonical `workspace.entrypoints` path.
 *
 * 11.47 B, in order, first match winning: exact path; bare basename when the
 * target contains no `/`; then the same two ASCII-case-folded, which emits the
 * same `config_warning` a case-folded `file:` does. Anything else raises
 * `unknown_pipeline` with the sorted entrypoints as candidates — including the
 * EMPTY target, which falls through the grammar to here because only the
 * document knows what the entrypoints are.
 */
export function resolveEntrypoint(
  entrypoints: string[],
  target: string,
): { entrypoint: string; warnings: string[] } {
  const bare = target.indexOf('/') < 0;
  for (const entry of entrypoints) {
    if (entry === target) return { entrypoint: entry, warnings: [] };
  }
  if (bare && target) {
    for (const entry of entrypoints) {
      if (basename(entry) === target) return { entrypoint: entry, warnings: [] };
    }
  }
  const want = asciiLower(target);
  const folded = (name: string): string[] => [
    'scope pipeline:' + target + ' matched case-insensitively; the canonical spelling is ' + name,
  ];
  if (target) {
    for (const entry of entrypoints) {
      if (asciiLower(entry) === want) return { entrypoint: entry, warnings: folded(entry) };
    }
    if (bare) {
      for (const entry of entrypoints) {
        if (asciiLower(basename(entry)) === want) return { entrypoint: entry, warnings: folded(entry) };
      }
    }
  }
  throw new ScopeError('unknown_pipeline', target, entrypoints);
}

/**
 * 11.47 C1 — the scope note, which exists to say what the view is NOT showing.
 *
 * Prose is free (11.2 step 10) and this is deliberately three numbers rather
 * than a sentence about one: a pipeline view that quietly dropped a shared
 * `data/` module, or that showed 40 of 300 nodes because 260 belong to no
 * pipeline at all, must not read as a clean bill of health.
 */
export function pipelineNote(
  index: PipelineIndex,
  entrypoint: string,
  totalNodes: number,
): string {
  const reached = index.reach(entrypoint).length;
  const shared = index.sharedWith(entrypoint).length;
  return (
    'pipeline:' + entrypoint + ' reaches ' + reached + ' of ' + totalNodes + ' node(s): ' + shared +
    ' shared with another entrypoint and drawn as context rather than claimed by this pipeline, and ' +
    index.unreached.length + ' node(s) of the whole graph belong to no pipeline at all. ' +
    'Only data and call edges join a pipeline — a config-only dependency is not shown here.'
  );
}

/**
 * Does the emitted `pipelines[]` block agree with the relation computed here?
 *
 * 11.47 A says neither port trusts the block, so nothing DEPENDS on this — but a
 * silent disagreement between the menu and the diagram is exactly the class of
 * bug this product cannot afford, so the chooser states it when it happens.
 * Returns null when there is nothing to say.
 */
export function pipelineDrift(graph: MLGraph, rows: PipelineRow[]): string | null {
  const block = (graph as { pipelines?: unknown }).pipelines;
  if (!Array.isArray(block)) return null;
  const emitted = block
    .map((row) => (row && typeof row === 'object' ? String((row as { entrypoint?: unknown }).entrypoint || '') : ''))
    .filter((name) => !!name);
  const computed = rows.map((r) => r.entrypoint);
  if (emitted.length === computed.length && emitted.every((name, i) => name === computed[i])) return null;
  return (
    'This document lists ' + emitted.length + ' pipeline(s) and the relation computed here finds ' +
    computed.length + '. The diagram follows the computed one; the difference usually means the ' +
    'document was capped or scoped after its block was written.'
  );
}

/**
 * MLV-P12 (11.47 B and C) — resolve one `pipeline:` selector into the shape
 * `scope/project.ts` step 2 hands to steps 3-11.
 *
 * It lives HERE rather than in `project.ts` because it is 11.47's rule, not
 * 11.2's: it replaces step 2 outright, it is the only caller of the relation
 * above, and `project.ts` was over the file-size bar with it inside.
 *
 * It replaces step 2 rather than extending it. `core` is what this entrypoint
 * reaches AND NOBODY ELSE DOES; what two entrypoints both reach becomes FORCED
 * CONTEXT, so a shared `data/` module is drawn as the frame it is rather than
 * claimed by whichever pipeline the reader happened to open. `anchors` stay the
 * SEEDS — the nodes actually written in the entrypoint file — because that is
 * what `view.resolvedTo` is for: "what did this selector name", not "what did it
 * pull in".
 */
export function resolvePipelineScope(
  graph: MLGraph,
  scope: Scope,
  nodes: MLNode[],
  warnings: string[],
): ScopeResolution {
  const entrypoints = (graph.workspace && graph.workspace.entrypoints) || [];
  const found = resolveEntrypoint(entrypoints, scope.target);
  for (const w of found.warnings) warnings.push(w);
  const index = new PipelineIndex(graph);
  const core = index.exclusive(found.entrypoint);
  const forcedContext = index.sharedWith(found.entrypoint);
  const anchors = index.seeds(found.entrypoint);
  // 11.47 C1. The note exists so a view that drew 40 of 300 nodes can never read
  // as a clean bill of health, and the two numbers that say so — how much is
  // shared, and how much belongs to no pipeline at all — are knowable only here.
  // It is appended only when the scope resolved to something, because an EMPTY
  // scope already gets 11.2 step 10's own "matched no nodes" note and two notes
  // about the same nothing is one too many. `core/project.py` gates it the same
  // way, on the same condition.
  if (anchors.length) warnings.push(pipelineNote(index, found.entrypoint, nodes.length));
  return {
    // The scope handed in, NOT a canonicalized copy: `view.scope` reports what
    // the user typed, exactly as `file:` does, and `core/project.py` builds its
    // `ScopeResolution` from `scope` too. Only the RELATION uses the canonical
    // entrypoint, which is what makes `pipeline:TRAIN.PY` draw the same nodes.
    scope,
    anchors,
    core,
    ambiguous: false,
    warnings,
    // 11.47 B: an entrypoint that resolves but whose file contributed no node —
    // it was capped away, or projected away — is an EMPTY SCOPE, not an error,
    // exactly as `unit:` is. `empty` is `not anchors` for every kind.
    empty: !anchors.length,
    forcedContext,
  };
}
